lib/utils/call.jl

```diff
@@ -150,3 +160,102 @@ end
+
+
+## version of ccall that calls jl_gc_safe_enter|leave around the inner ccall
+
+# TODO: replace with JuliaLang/julia#49933 once merged
+
+# note that this is generally only safe with functions that do not call back into Julia.
+# when callbacks occur, the code should ensure the GC is not running by wrapping the code
+# in the `@gcunsafe` macro
+
+function ccall_macro_lower(func, rettype, types, args, nreq)
+    # instead of re-using ccall or Expr(:foreigncall) to perform argument conversion,
+    # we need to do so ourselves in order to insert a jl_gc_safe_enter|leave
+    # just around the inner ccall
+
+    cconvert_exprs = []
+    cconvert_args = []
+    for (typ, arg) in zip(types, args)
+        var = gensym("$(func)_cconvert")
+        push!(cconvert_args, var)
+        push!(cconvert_exprs, quote
+            $var = Base.cconvert($(esc(typ)), $(esc(arg)))
+        end)
+    end
+
+    unsafe_convert_exprs = []
+    unsafe_convert_args = []
+    for (typ, arg) in zip(types, cconvert_args)
+        var = gensym("$(func)_unsafe_convert")
+        push!(unsafe_convert_args, var)
+        push!(unsafe_convert_exprs, quote
+            $var = Base.unsafe_convert($(esc(typ)), $arg)
+        end)
+    end
+
+    call = quote
+        $(unsafe_convert_exprs...)
+
+        gc_state = @ccall(jl_gc_safe_enter()::Int8)
+        ret = ccall($(esc(func)), $(esc(rettype)), $(Expr(:tuple, map(esc, types)...)),
+                    $(unsafe_convert_args...))
+        @ccall(jl_gc_safe_leave(gc_state::Int8)::Cvoid)
+        ret
+    end
+
+    quote
+        $(cconvert_exprs...)
+
+        GC.@preserve $(cconvert_args...)  $(call)
+    end
+end
+
+"""
+    @gcsafe_ccall ...
+
+Call a foreign function just like `@ccall`, but marking it safe for the GC to run. This is
+useful for functions that may block, so that the GC isn't blocked from running, but may also
+be required to prevent deadlocks (see JuliaGPU/CUDA.jl#2261).
+
+Note that this is generally only safe with non-Julia C functions that do not call back into
+Julia. When using callbacks, the code should make sure to transition back into GC-unsafe
+mode using the `@gcunsafe` macro.
+"""
+macro gcsafe_ccall(expr)
+    ccall_macro_lower(Base.ccall_macro_parse(expr)...)
+end
+
+"""
+    @gcunsafe_callback function callback(...)
+        ...
+    end
+
+Mark a callback function as unsafe for the GC to run. This is normally the default for
+Julia code, and is meant to be used in combination with `@gcsafe_ccall`.
+"""
+macro gcunsafe_callback(ex)
+    if VERSION >= v"1.9"
+        # on 1.9+, `@cfunction` already transitions to GC-unsafe mode
+        return esc(ex)
+    end
+
+    # parse the function definition
+    @assert Meta.isexpr(ex, :function)
+    sig = ex.args[1]
+    @assert Meta.isexpr(sig, :call)
+    body = ex.args[2]
+    @assert Meta.isexpr(body, :block)
+
+    gcunsafe_body = quote
+        gc_state = @ccall(jl_gc_unsafe_enter()::Int8)
+        try
+            $body
+        finally
```

Comment:
I was debating if the try/finally was necessary and decided that it wasn't. Since we can only catch exceptions in gc_unsafe. cc: @vtjnash for a double check on that.

Reply:
The `try` setup seems a bit risky to be present inside `jl_gc_unsafe_enter`, a `catch/finally` already implicitly reset the gc state upon leave (normal or error)

Reply:
The `try` happens after `jl_gc_unsafe_enter`?

Reply:
Then I don't see how this is reachable, since cfunction already had to do an unsafe-enter to be able to dispatch into the runtime

Reply:
Oh, I see this is for v1.8. That seems okay. I am surprised that is still supported here